extensions/smallrye-metrics/deployment/src/main/java/io/quarkus/smallrye/metrics/deployment/SmallRyeMetricsProcessor.java

```diff
@@ -165,28 +150,39 @@ public void logCleanup(BuildProducer<LogCleanupFilterBuildItem> logCleanupFilter
     void registerMetricsFromAnnotatedMethods(SmallRyeMetricsRecorder metrics,
             BeanContainerBuildItem beanContainerBuildItem,
             BeanArchiveIndexBuildItem beanArchiveIndex) {
-        JandexBeanInfoAdapter beanInfoAdapter = new JandexBeanInfoAdapter(beanArchiveIndex.getIndex());
-        JandexMemberInfoAdapter memberInfoAdapter = new JandexMemberInfoAdapter(beanArchiveIndex.getIndex());
-        for (ClassInfo clazz : beanArchiveIndex.getIndex().getKnownClasses()) {
-            // TODO: maybe leave out more cases where we can be sure that there are no metrics to be registered
-            // to speed up the startup
-            if (clazz.name().prefix().toString().startsWith("io.smallrye.metrics")) {
-                continue;
+        IndexView index = beanArchiveIndex.getIndex();
+        JandexBeanInfoAdapter beanInfoAdapter = new JandexBeanInfoAdapter(index);
+        JandexMemberInfoAdapter memberInfoAdapter = new JandexMemberInfoAdapter(index);
+
+        Map<DotName, ClassInfo> collectedMetricsClasses = new HashMap<>();
+
+        for (DotName metricAnnotation : METRICS_ANNOTATIONS) {
+            Collection<AnnotationInstance> metricAnnotationInstances = index.getAnnotations(metricAnnotation);
+            for (AnnotationInstance metricAnnotationInstance : metricAnnotationInstances) {
+                AnnotationTarget metricAnnotationTarget = metricAnnotationInstance.target();
+                switch (metricAnnotationTarget.kind()) {
+                    case METHOD: {
+                        ClassInfo clazz = metricAnnotationTarget.asMethod().declaringClass();
+                        collectedMetricsClasses.put(clazz.name(), clazz);
+                        break;
+                    }
+                    case CLASS: {
+                        ClassInfo clazz = metricAnnotationTarget.asClass();
+                        collectedMetricsClasses.put(clazz.name(), clazz);
+                        break;
+                    }
+                }
             }
+        }
+
+        for (ClassInfo clazz : collectedMetricsClasses.values()) {
```

Comment:
There's something weird here: if the type is not registered but only some methods are registered, then we should not register metrics for all the methods of the classes.

Or did I miss something?

Reply:
> ..if the type is not registered but only some methods are registered, then we should not register metrics for all the methods of the classes.

I think this is done in `MetricResolver.class` 
```
 private <T extends Annotation> MetricResolver.Of<T> beanResolverOf(MemberInfo element, Class<T> metric, BeanInfo bean) {
        if (bean.isAnnotationPresent(metric)) {
            AnnotationInfo annotation = bean.getAnnotation(metric);
            String name = this.metricName(bean, element, metric, annotation.name(), annotation.absolute());
            Tag[] tags = this.metricTags(annotation);
            return new MetricResolver.DoesHaveMetric(annotation, name, tags);
        } else {
            return (MetricResolver.Of)(bean.getSuperclass() != null ? this.beanResolverOf(element, metric, bean.getSuperclass()) : new MetricResolver.DoesNotHaveMetric());
        }
    }
```

Reply:
I'm not sure I have been clear here.

Right now, we do:
- if method annotated, add the class to a set (or whatever structure)
- if class annotated, add the class to a set
- for each class, register metrics for all the methods

Wouldn't it be better to do:
- if method annotated, add the method to a set
- if class annotated, add the class to a set
- for each method which is not part of a registered class, register metrics for the method
- for each class, register metrics for all the methods of the class

The code you point out makes me think we should also consider all the subclasses of the annotated classes (only when the annotation is defined on the class AFAICS).

Reply:
Thanks for the clarification. 

My understanding was that the above logic was handled in the resolver code I pointed above and in `Metadata.registerMetrics()` which checks if any metric is present before registering a particular metric. 

But maybe it is better to explicit this logic in the processor as you suggested. 
 
> The code you point out makes me think we should also consider all the subclasses of the annotated classes (only when the annotation is defined on the class AFAICS).

+1 good catch.

Reply:
Suggestion taken into account. WDYT @gsmet ?